dataplane/src/nat.rs

```diff
@@ -19,88 +27,60 @@ struct Pif {
     vpc: String,                  // Name of the associated VPC
 }
 
-#[derive(Default, Debug, Clone)]
-struct TrieNode<K, V>
-where
-    K: Clone + Eq + Hash,
-    V: Clone,
-{
-    children: HashMap<K, TrieNode<K, V>>,
-    value: Option<V>,
-}
-
-#[derive(Default, Debug, Clone)]
-struct PrefixTrie<K, V>
-where
-    K: Clone + Eq + Hash,
-    V: Clone,
-{
-    root: TrieNode<K, V>,
+#[derive(Default, Clone)]
+struct PrefixTrie {
+    trie_ipv4: RTrieMap<Ipv4Prefix, String>,
+    trie_ipv6: RTrieMap<Ipv6Prefix, String>,
 }
 
-impl<K, V> PrefixTrie<K, V>
-where
-    K: Clone + Eq + Hash + Default + Debug,
-    V: Clone + Default + Debug,
-{
+impl PrefixTrie {
     #[tracing::instrument(level = "trace")]
     fn new() -> Self {
         Self {
-            root: TrieNode::default(),
+            trie_ipv4: RTrieMap::new(),
+            trie_ipv6: RTrieMap::new(),
         }
     }
 
     #[tracing::instrument(level = "trace")]
-    fn insert(&mut self, keys: Vec<K>, value: V) -> bool {
-        let mut node = &mut self.root;
-
-        for key in keys {
-            node = node.children.entry(key).or_default();
-        }
-
-        if node.value.is_some() {
-            return false; // Prefix already exists
-        }
-
-        node.value = Some(value);
-        true
-    }
-
-    #[tracing::instrument(level = "trace")]
-    fn find(&self, keys: Vec<K>) -> Option<V> {
-        let mut node = &self.root;
-        let mut best_match = None;
-
-        for key in keys {
-            if let Some(val) = &node.value {
-                best_match = Some(val.clone());
+    fn insert(&mut self, addr: &IpAddr, len: &u8, value: String) -> Result<(), NatError> {
+        match addr {
+            IpAddr::V4(ip) => {
+                let prefix = Ipv4Prefix::new(*ip, *len).or(Err(NatError::BadPrefix))?;
+                if self.trie_ipv4.insert(prefix, value).is_some() {
```

Comment:
Hum. I am not sure this is correct (or what you intend to do).
IIRC, trie.insert() will insert the new value (string) and return the previous one, if there was any.
With this code, an insert of value=BAR would overwrite value=FOO (if the value existed for the same prefix) and then the code would complain that it existed already. But the end value for the prefix would be BAR, not FOO.... right?

Reply:
I think you could do something like (untested):
```
if self.trie_ipv4.get(prefix, value).is_some() {
     Err(NatError::PifExists);
} else {
    self.trie_ipv4.insert(prefix, value);
    Ok(())
}
```

Reply:
Yes, correct. I realised that I needed to fix it at some point last night, but eventually forgot about it :facepalm:. I'll address it, thank you.